### PR TITLE
Add pod annotations to container

### PR DESCRIPTION
**Pull request by haircommander**
We have the annotations SandboxID and SandboxName, let's use them. This also allows kata containers to be created in pods

partial fix to https://github.com/containers/libpod/issues/4353 (we don't need to add the annotations manually, but still need to disable selinux labeling manually) 

Signed-off-by: Peter Hunt <pehunt@redhat.com>

### CODE REVIEW

**Comment by haircommander** [comment]
PTAL @zer0def

**Comment by TomSweeneyRedHat** [comment]
LGTM, assuming that's the right spot to retain the info.  I'll let other chime in on that part.

File: cmd/podman/shared/create.go [hunk 1]
```diff
@@ -322,7 +322,7 @@ func configurePod(c *GenericCLIResults, runtime *libpod.Runtime, namespaces map[
 	if (namespaces["uts"] == cc.Pod) || (!c.IsSet("uts") && pod.SharesUTS()) {
 		namespaces["uts"] = fmt.Sprintf("container:%s", podInfraID)
 	}
-	return namespaces, nil
+	return namespaces, pod.ID(), nil
```
Comment (zer0def) [inline cmd/podman/shared/create.go]: This should be either of:
- infra container's ID (only then `/var/lib/containers/storage/overlay-containers/${INFRA_CONTAINER_ID}/userdata/config.json` exists for other containers in the same pod to latch onto)
- if we don't have an infra container, the ID of whichever container is running with ContainerType == `sandbox` (there should always be only one sandbox container within a pod, but the same condition applies)

Either of these containers has to be running in order to be able to attach other containers with ContainerType == `container` into the sandbox.

Comment (haircommander) [reply cmd/podman/shared/create.go]: From the sounds of it @zer0def, it seems like we should only allow this kind of sharing if the sandbox container is an infra container. There is already mechanisms in place to guarentee the infra container isn't deleted before other containers in the pod, which seems like something that could effect kata.

For instance, if the first pod in the container is created, another is added (and attached to its vm) and the first is deleted, I'm pretty sure that container's config.json is going to be removed. Would that negatively effect the other containers in the pod?

**Comment by haircommander** [comment]
/hold
this needs some thinking

File: cmd/podman/shared/create.go (hunk 1 above)
Comment (zer0def) [reply cmd/podman/shared/create.go]: Indeed, the intended idea would be to have the infra container be the pod's sandbox container, but I wouldn't immediately dismiss other ideas, if there's sufficient motivation for doing so with careful execution.

Comment (haircommander) [reply cmd/podman/shared/create.go]: okay, I'll stick to an infra container for now

**Comment by zer0def** [comment]
As of `1342d713dc6bd2aca15039f323ae87a99abfda98` along with the patch at the end of this message:
- flow from #4353 point 2.2. works, even if it causes the sandbox/VM to SIGABRT on pod destruction (which might be a separate issue outside of the scope of this one)
- flow from 2.1 of the same issue fails without explicitly stating the annotations that would attach the container to a given sandbox, plus given how this was used as a workaround to just get *any* working result, I'd say it's fine to leave it be as-is
- `play kube` still fails along the same lines as 2.2 did, but that should be an easy port

I've also removed the complementing condition for `podInfraID != ""` that explicitly set `ContainerType=sandbox` just to err on the safe side in case there's ever a runtime whose behavior would conflict with this annotation in some way. After testing, it worked just fine.

```
diff --git a/cmd/podman/shared/create.go b/cmd/podman/shared/create.go
index b9c1aa86..e158a8c2 100644
--- a/cmd/podman/shared/create.go
+++ b/cmd/podman/shared/create.go
@@ -12,7 +12,7 @@ import (
 	"syscall"
 	"time"
 
-	"github.com/containers/image/v4/manifest"
+	"github.com/containers/image/v5/manifest"
 	"github.com/containers/libpod/cmd/podman/shared/parse"
 	"github.com/containers/libpod/libpod"
 	"github.com/containers/libpod/libpod/image"
@@ -586,8 +586,6 @@ func ParseCreateOpts(ctx context.Context, c *GenericCLIResults, runtime *libpod.
 	if podInfraID != "" {
 		annotations[ann.SandboxID] = podInfraID
 		annotations[ann.ContainerType] = ann.ContainerTypeContainer
-	} else {
-		annotations[ann.ContainerType] = ann.ContainerTypeSandbox
 	}
 
 	if data != nil {
```

**Comment by rhatdan** [comment]
@haircommander Any update on this PR?

**Comment by haircommander** [comment]
@rhatdan @zer0def whoops, this fell through the cracks. Added the changes as suggested. PTAL

**Comment by rhatdan** [comment]
@zer0def could we make it work with play kube?

**Comment by zer0def** [comment]
@rhatdan I'm not much of a golang coder, but since current changes revolve around `cmd/podman/shared/create.go:ParseCreateOpts` that returns a `CreateConfig`, my guess would be that similar logic could be introduced to `pkg/adapter/pods.go:kubeContainerToCreateConfig`. It would require confirmation from someone a bit better versed with the language and codebase than I current am, though.

**Comment by haircommander** [comment]
I can take care of that later today

**Comment by zer0def** [comment]
Something like this does the trick, though I would deem it more of a hack than a proper solution, with code duplication and all.

```
diff --git a/pkg/adapter/pods.go b/pkg/adapter/pods.go
index 6648edc8..ae118c79 100644
--- a/pkg/adapter/pods.go
+++ b/pkg/adapter/pods.go
@@ -17,6 +17,7 @@ import (
        "github.com/containers/libpod/libpod"
        "github.com/containers/libpod/libpod/image"
        "github.com/containers/libpod/pkg/adapter/shortcuts"
+       ann "github.com/containers/libpod/pkg/annotations"
        ns "github.com/containers/libpod/pkg/namespaces"
        createconfig "github.com/containers/libpod/pkg/spec"
        "github.com/containers/libpod/pkg/util"
@@ -791,6 +792,19 @@ func kubeContainerToCreateConfig(ctx context.Context, containerYAML v1.Container
        }
        containerConfig.Env = envs
 
+       annotations := make(map[string]string)
+       pod, err := runtime.LookupPod(podID)
+       if err != nil {
+               return nil, err
+       }
+       podInfraID, err := pod.InfraContainerID()
+       if err != nil {
+               return nil, err
+       }
+       annotations[ann.SandboxID] = podInfraID
+       annotations[ann.ContainerType] = ann.ContainerTypeContainer
+       containerConfig.Annotations = annotations
+
        for _, volume := range containerYAML.VolumeMounts {
                hostPath, exists := volumes[volume.Name]
                if !exists {
```

**Comment by haircommander** [comment]
@zer0def thanks, your approach was basically what I was going to do :)

**Comment by zer0def** [comment]
Can confirm both cases of creating a pod with an infra container and `play kube` run just fine with this PR.

**Comment by amshinde** [comment]
@haircommander Thanks for the PR.
I verified that the containers for a pod are launched in a single qemu instance for Kata with this PR provided that the infra container is present.
Afaik, Kata should not have a dependency for the infra container, (we had one earlier for sharing PID namespaces)
So, I wanted to try to see if that would work with podman, but what I am seeing with podman is that without the infra container, containers in a pod have different IP addresses and hostname, which surprised me. 
Aren't containers in a pod supposed to have the same IP address and share network,uts and ipc namespaces?

cc @giuseppe

**Comment by haircommander** [comment]
@amshinde right now, we're kind of using the infra container as a crutch (mostly because I don't really understand the needs of kata).

Let's imagine pods and kata work without infra. if a pod is created without an infra container, the first container would become the owner of the vm (annotated as sandbox) and the other containers that are added to the pod attach to that vm (annotated as container). 

If the first container in the pod is removed (which it can be if it's not marked an infra), what happens to that VM? who should the other containers add as a reference for `sandboxID` (to attach to the VM)

Relying on the infra container here is the most clear way to make sure there is a container running that holds the VM, that the other containers can always rely on joining to. 

is my understanding correct for how the kata VM works?

**Comment by amshinde** [comment]
> Let's imagine pods and kata work without infra. if a pod is created without an infra container, the first container would become the owner of the vm (annotated as sandbox) and the other containers that are added to the pod attach to that vm (annotated as container).

You are right there. But Kata also does not rely on the pause container being the last one to exit iirc.
So, I thought this could work. (The `sandboxId` could be the `id` of the first container or a uniquely generated one)
But given the limitation of pods without infra container in podman today(https://github.com/containers/libpod/issues/4531), this would definitely not work with Kata. Kata does rely on containers sharing the same namespaces and having a single IP.

**Comment by zer0def** [comment]
@amshinde You can still create a pod without an infra container (which is also what would be, as you put it, "a uniquely generated one") by following [these instructions](https://github.com/containers/libpod/issues/4353#issuecomment-547159196) coutresy of @mheon . It's just unnecessarily tideous, on top of not providing namespace guarantees for other containers in the pod to rely on, as @haircommander outlined.

Do mind that this PR hasn't yet been part of a stable release, in which case the behavior you're describing is expected as of the current latest stable, because each container will be started as a separate VM due to not receiving appropriate container-type and sandbox-id annotations.

**Pull request merged**